ood-portal-generator/lib/ood_portal_generator/application.rb

```diff
@@ -29,56 +33,246 @@ def context
+        $?.success?
+      end
+
+      def exit!(code)
+        case code
+        when TrueClass
+          exit(0)
+        when FalseClass
+          exit(1)
+        else
+          exit(code)
+        end
+      end
+
+      def generate()
+        view = View.new(context)
+        rendered_template = view.render(template.read)
+        output.write(rendered_template)
+      end
+
+      def update_ood_portal()
+        ret = 0
+        changed = false
+        new_apache = Tempfile.new('new_apache')
+        @output = new_apache.path
+        generate()
+
+        # Create checksum file if the path to ood-portal.conf not in checksum file
+        # Checksum is based on mktemp generated ood-portal.conf but using path of real ood-portal.conf
+        if ! checksum_exists?
+          puts "Generating Apache config checksum file: '#{sum_path}'"
+          save_checksum(new_apache.path)
+        end
+
+        replace = update_replace?
+
+        if ! files_identical?(new_apache.path, apache)
+          if replace
+            if File.exist?(apache)
+              puts "Backing up previous Apache config to: '#{apache_bak}'"
+              FileUtils.mv(apache, apache_bak)
+            end
+            puts "Generating new Apache config at: '#{apache}'"
+            `cat "#{new_apache.path}" > "#{apache}"`
+            puts "Generating Apache config checksum file: '#{sum_path}'"
+            save_checksum(apache)
+            ret = change_exit
+            changed = true
+          else
+            puts "WARNING: Checksum of #{apache} does not match previous value, not replacing."
+            puts "Generating new Apache config at: '#{apache}.new'"
+            `cat "#{new_apache.path}" > "#{apache}.new"`
+            ret = skip_exit
+          end
+        else
+          puts "No change in Apache config."
+        end
+
+        new_apache.unlink
+
+        if rpm || ! replace
+          return ret
+        end
+
+        puts "Completed successfully!"
+        if changed
+          puts ""
+          puts "Restart the #{apache_services[0]} service now."
+          puts ""
+          puts "Suggested command:"
+          puts "    sudo systemctl try-restart #{apache_services[0]}.service #{apache_services[1]}.service"
+          puts ""
+        end
+
+        return ret
+      end
+
       # Starts the OodPortalGenerator CLI
       # @return [void]
-      def start
+      def start(mode, argv = ARGV)
         # Set a cleaner process title
-        Process.setproctitle("generate #{ARGV.join(" ")}")
+        Process.setproctitle("#{mode} #{argv.join(" ")}")
 
         # Parse CLI arguments
-        parser.parse!
+        OptionParser.new do |parser|
+          parser.banner = "Usage: #{mode} [options]"
+
+          if mode == 'generate'
+            add_generate_opt_parser_attrs(parser)
+            default = "-c #{config} -t #{template}"
+          elsif mode == 'update_ood_portal'
+            add_generate_opt_parser_attrs(parser, false)
```

Comment:
Since `add_generate_opt_parser_attrs` is called in both, is there a reason why we wouldn't move the options shared between both to `add_shared_opt_parser_attrs`? If we did that, then `add_generate_opt_parser_attrs` would only have the output option....

Reply:
Ya that makes sense, made the change.